[IA-1756] PD data model

rtitle opened the pull request:
Data model, slick mapping, and liquibase migration for PDs.

See:
https://broadworkbench.atlassian.net/browse/IA-1756
https://broadworkbench.atlassian.net/wiki/spaces/IA/pages/398295082/2020-04-22+Data+Model+for+Persistent+Disks

---
Have you read [CONTRIBUTING.md](https://github.com/DataBiosphere/leonardo/blob/develop/CONTRIBUTING.md) lately? If not, do that first.

I, the developer opening this PR, do solemnly pinky swear that:

- [ ] I've documented my API changes in Swagger

In all cases:

- [ ] Get a thumbsworth of review and PO signoff if necessary
- [ ] Verify all tests go green
- [ ] Squash and merge; Delete your branch after this
- [ ] Test this change deployed correctly and works on dev environment after deployment

rtitle committed:
Models, codecs, and slick mapping

rtitle committed:
Add migration and unit tests

rtitle committed:
Fix unit tests

Qi77Qi commented on core/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/JsonCodec.scala:
```diff
@@ -111,6 +112,14 @@ object JsonCodec {
     "timestamp"
   )(x => RuntimeError.unapply(x).get)
+  implicit val diskStatusEncoder: Encoder[DiskStatus] = Encoder.encodeString.contramap(_.toString)
+  implicit val diskAuditInfoEncoder: Encoder[DiskAuditInfo] = Encoder.forProduct4(
```
this looks about the same as existing AuditInfo, can we reuse the same case class?

rtitle replied:
That would be nice… I noticed that the existing `AuditInfo` had 1 field that's not relevant to disks. Will see if I can refactor so we can reuse it.

jdcanas replied:
I'm running into the same thing with kubernetes right now. I think creator, createdDate, destroyedDate are the core fields between all 3 use cases

Qi77Qi commented on core/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/diskModels.scala:
```diff
@@ -1,6 +1,62 @@
+import org.broadinstitute.dsde.workbench.model.WorkbenchEmail
+import org.broadinstitute.dsde.workbench.model.google.GoogleProject
+
+final case class PersistentDisk(id: Long,
```
can we use a `DiskId` type, just so we will never mix up runtime id with this?

Qi77Qi commented on http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/PersistentDiskComponent.scala:
```diff
@@ -0,0 +1,150 @@
+                                      creator: WorkbenchEmail,
+                                      createdDate: Instant,
+                                      destroyedDate: Instant,
+                                      dateAccessed: Instant,
```
can we move these fields into `AuditInfo`?

rtitle replied:
We could.. but this lets us write `<> (PersistentDiskRecord.tupled, PersistentDiskRecord.unapply)` below. I kind of see this class as a direct mapping to columns in the database, whereas the `PersistentDisk` class has more structure.

Qi77Qi replied:
I don't feel strongly either way..But if we see this as direct mapping, it's easy to run into 22 fields limit tho.

Qi77Qi commented on http/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/CommonTestData.scala:
```diff
@@ -287,6 +291,21 @@ object CommonTestData {
   def modifyInstanceKey(instanceKey: DataprocInstanceKey): DataprocInstanceKey =
     instanceKey.copy(name = InstanceName(instanceKey.name.value + "_2"))
+
+  def makePersistentDisk(id: Int): PersistentDisk = PersistentDisk(
```
shall we start making scalacheck generators instead?

rtitle replied:
Hmm.. yeah maybe

Qi77Qi reviewed (commented):
👍 lgtm in general..left some minor comments

jdcanas commented on http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/PersistentDiskComponent.scala:
```diff
@@ -0,0 +1,150 @@
+                                      zone: ZoneName,
+                                      name: DiskName,
+                                      googleId: Option[GoogleId],
+                                      samResourceId: DiskSamResourceId,
```
why do we use this here, but not do the same for GCE runtimes/Dataproc clusters?

rtitle replied:
You mean `DiskSamResourceId`? Runtimes have a field for this but it's called `RuntimeInternalId` which IMO is a bit non-descriptive because it's only used for Sam IDs.

Qi77Qi replied:
I'm wondering, is it reasonable to make a `SamResourceId` data type and use it for all different sam resources

rtitle replied:
I thought about that, was thinking it's good to have type safety to not confuse runtime and disk sam IDs. 

Maybe we could add an ADT like:
```
sealed trait SamResource {
  def resourceId: String
  def resourceType: SamResourceType
}
object SamResource {
  final case class RuntimeSamResource(resourceId: String) {
    override def resourceType: SamResourceType.Runtime
  }
  final case class PersistentDiskSamResource(resourceId: String) {
    override def resourceType: SamResourceType.PersistentDisk
  }
}
```

rtitle replied:
So I started down this path and it led to a lot of other refactoring… I still think it's valuable but I think I will split it into a separate PR, so this can be merged sooner. Will still apply the other feedback in this PR.

rtitle replied:
Opened https://broadworkbench.atlassian.net/browse/IA-1893

rtitle commented on http/src/main/resources/org/broadinstitute/dsde/workbench/leonardo/liquibase/changesets/20200427_create_persistent_disk_table.xml:
```diff
@@ -0,0 +1,71 @@
+        </createIndex>
+    </changeSet>
+    <changeSet logicalFilePath="leonardo" author="rtitle" id="persistent-disk-label-table">
+        <createTable tableName="PERSISTENT_DISK_LABEL">
```
Thinking more, I think I'm going to try and repurpose the existing `LABEL` table instead of adding a new `PERSISTENT_DISK_LABEL`

jdcanas replied:
➕ I was starting to write a comment along the lines of this

akarukappadath replied:
If disks can get errored out, I wonder if we should do the same for the `CLUSTER_ERROR` table

rtitle replied:
That's an excellent question… will think about it but probably punt to a separate PR if so.

jdcanas commented on http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/PersistentDiskComponent.scala:
```diff
@@ -0,0 +1,150 @@
+                                      diskType: DiskType,
+                                      blockSize: BlockSize)
+
+class PersistentDiskTable(tag: Tag) extends Table[PersistentDiskRecord](tag, "PERSISTENT_DISK") {
```
I think you need to define an `index` to correspond to `IDX_PERSISTENT_DISK_UNIQUE`

rtitle replied:
We typically haven't represented all our indexes in the slick mapping code. The indexes are still present in the database. IIRC they are only useful in slick when generating the schema from slick code, which we don't do

rtitle committed:
1. Rebase
2. PR feedback: re-use AuditInfo in PersistentDisk model
3. PD feedback: use DiskId type in PersistentDisk model
Not done yet: LABEL table refactor

akarukappadath commented on core/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/diskModels.scala:
```diff
@@ -1,6 +1,55 @@
+                                size: DiskSize,
+                                diskType: DiskType,
+                                blockSize: BlockSize,
+                                labels: LabelMap)
```
Since we have a `Failed` disk status, should we have an `errors` field here?

rtitle replied:
Yeah, I didn't make `errors` at all.. was thinking of punting that till we have more understanding of what errors can occur.

akarukappadath reviewed (approved):
Looks good to me - just a question about PD errors

Qi77Qi reviewed (approved):
lgtm

rtitle committed:
Make label table polymorphic

rtitle commented on core/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/runtimeModels.scala:
```diff
@@ -25,6 +25,7 @@ final case class Runtime(id: Long,
                          serviceAccount: WorkbenchEmail,
                          asyncRuntimeFields: Option[AsyncRuntimeFields],
                          auditInfo: AuditInfo,
+                         kernelFoundBusyDate: Option[Instant],
```
Moved `kernelFoundBusyDate` out of `AuditInfo` and into `Runtime`
